### DISCUSSION TITLE
Issue with at.load_madx conversion

### DISCUSSION THREAD

**Discussion by elafmusa** [atcollab/at#850] [Python AT]
Hi everyone,

I am trying to use at.load_madx to convert a MADX sequence file [(available here)](https://syncandshare.desy.de/index.php/s/WsRXr5jqmjLynfm) as follows:

```
ring = load_madx("z_ballistic.seq", use="fccee_p_ring", energy = 45.6e9)
ring.enable_6d()
print(ring.envelope_parameters())
```

However, I get the error from the envelope parameters print: array must not contain infs or NaNs. Additionally, during conversion, I see the message: "MADX type Collimator is ignored and replaced by a Drift."

As an alternative, I used xconverters to convert the same file, and it worked without errors:
```

from xconverters import convert_lattices
xlat = convert_lattices.from_madx_seqfile('z_ballistic.seq', seq_name='fccee_p_ring', particle='electron', energy=45.6, dependencies=False)
ring = convert_lattices.to_pyat(xlat)

ring.enable_6d()
print(ring.envelope_parameters())
```

 Frac. tunes (6D motion): [0.75791762 0.37388092 0.02895448]
                   Energy: 4.560000e+10 eV
       Energy loss / turn: 3.908958e+07 eV
          Mode emittances: [ 8.56422504e-10 -1.44757890e-37  2.18746556e-06]
Damping partition numbers: [1.00045508 1.00001076 1.99953415]
            Damping times: [0.70521964 0.70553298 0.35285247] s
            Energy spread: 0.000391665
             Bunch length: 0.00558646 m
         Cavities voltage: 79200000.0 V
        Synchrotron phase: 2.62542 rd
    Synchrotron frequency: 95.7474 Hz
    
I would like to confirm that I am using at.load_madx correctly, or if the Collimator handling is causing this issue.

Thank you in advance,
Elaf

**Comment by lfarv** [comment]
Hello @elafmusa

Thanks for reporting the problem. Could you post your `z_balistic.seq` file, so that I can look at it ?

**Reply by elafmusa** [reply]
Dear @lfarv 

Thank you for your answer, please find the lattice file in this link:

https://syncandshare.desy.de/index.php/s/WsRXr5jqmjLynfm

**Comment by lfarv** [comment]
Thanks ! Could you also send the correctly translated lattice as a .mat file ? I’ll try to understand where the differences are with the result of load_madx…Le 2 nov. 2024 à 08:51, elafmusa ***@***.***> a écrit :﻿
Dear @lfarv
Thank you for your answer, please find the lattice file in this link:
https://syncandshare.desy.de/index.php/s/WsRXr5jqmjLynfm

**Comment by elafmusa** [comment]
Dear @lfarv ,

Please find the .mat file in this link:

https://syncandshare.desy.de/index.php/s/oS7rBWzXg5KZTjR

Best regards,
Elaf

**Comment by lfarv** [comment]
@elafmusa : there was indeed a problem in `load_madx`, which is corrected in #851.

After correction, there are still 3 small differences between the result of `load_madx` and your .mat file:

- the lattice in the .mat file has radiation in sextupoles. This is not the default behavior of `enable_6d`. Is this due to xconverters, or did you explicitly set it?
- `load_madx` keeps the rf frequency as it is specified in the RFCAVITY elements. This leads to an unstable machine. Setting it to the nominal frequency by calling `ring.set_rf_frequency()` brings it to exactly the same value as in the .mat file. Was the frequency set by xconverters ?
- xconverters adds two additional markers at the beginning and end of the lattice, `load_madx` does not.

Finally, after ensuring that radiating elements are the same, and that the rf frequency is the same, both lattices are identical. Here is how I got the two lattice:
```python
seqring=at.load_madx(seqname, use="fccee_p_ring", energy = 45.6e9)
seqring.enable_6d()         # enable radiation
seqring.set_rf_frequency()  # set rf frequency to nominal
```
and
```python
matring = at.load_mat(matname)[1:-1]                         # Remove the beginning and end markers
matring.enable_6d(sextupole_pass="StrMPoleSymplectic4Pass")  # remove radiation in sextupoles
```

Thanks for reporting the problem. I tried to use xconverters to check, but I get an error when importing `convert_lattices`…

**Comment by elafmusa** [comment]
Dear @lfarv,

Thank you very much!

As for the converted lattice using xconverters i run exactly the following lines:
```

from xconverters import convert_lattices
xlat = convert_lattices.from_madx_seqfile('z_ballistic.seq', seq_name='fccee_p_ring', particle='electron', energy=45.6, dependencies=False)
ring = convert_lattices.to_pyat(xlat)
ring.enable_6d(dipole_pass='BndMPoleSymplectic4RadPass',
            quadrupole_pass='StrMPoleSymplectic4RadPass',
            sextupole_pass='StrMPoleSymplectic4RadPass')
```

before saving the .mat file 

`ring.save('z_ballistic.mat', mat_key='ring')`

**Comment by lfarv** [comment]
@elafmusa
This answers two questions:
1. xconverters automatically tunes the rf frequency to the "nominal" value, `load_madx` does not. Should it do it ? Up to now, creating a lattice from any source takes the data as it is, with no modification. I would prefer `load_madx` to stick to that rule, unless there is a general agreement to change.
2. Radiation in the sextupoles is your choice, so nothing to do…

In the mean time, I found another problem converting entrance/exit angles of bending magnets with a negative bending angle. I'm working on that…

**Comment by lfarv** [comment]
Fixed in #851
